rfcs/20210513-pluggable-profiler-for-tensorflow.md

```diff
@@ -0,0 +1,429 @@
+  void (*stop)(const TP_Profiler* profiler, TF_Status* status);
+
+  // Saves collected profile data into XSpace and serializes it to the buffer.
+  // If this have been called, subsequent calls might
+  // return empty data.
+  void (*collect_data_xspace)(const TP_Profiler* profiler, uint8_t* buffer, size_t* size_in_bytes, TF_Status* status);
+} TP_ProfilerFns;
+
+#define TP_PROFILER_FNS_STRUCT_SIZE TF_OFFSET_OF_END(TP_ProfilerFns, collect_data_xspace)
+
+typedef struct TF_ProfilerRegistrationParams {
+  size_t struct_size;
+  void* ext; // reserved for future use
+
+  // TensorFlow Profiler C API version.
+  int32_t major_version;
+  int32_t minor_version;
+  int32_t patch_version;
+
+  [in/out] Memory owned by core but attributes within are populated by the plugin.
+  TP_Profiler* profiler;
+  [in/out] Memory owned by core but attributes within are populated by the plugin.
+  TP_ProfilerFns* profiler_fns;
+  // [out] Pointer to plugin's `TP_Profiler` clean up function. 
+  // Cleans up fields inside `TP_Profiler` that were allocated
+  // by the plugin. `profiler` itself must not be deleted by the plugin.
+  void (*destroy_profiler)(TP_Profiler* profiler);
+  // [out] Pointer to plugin's `TP_ProfilerFns` clean up function. 
+  // Cleans up fields inside `TP_ProfilerFns` that were allocated
+  // by the plugin. `profiler_fns` itself must not be deleted by the plugin.
+  void (*destroy_profiler_fns)(TP_ProfilerFns* profiler_fns);
+} TF_ProfilerRegistrationParams;
+
+#define TF_PROFILER_REGISTRATION_PARAMS_STRUCT_SIZE \
+  TF_OFFSET_OF_END(TF_ProfilerRegistrationParams, destroy_profiler_fns)
+
+void TF_InitProfiler(TF_ProfilerRegistrationParams* params, TF_Status* status);
+
+#ifdef __cplusplus
+} // extern "C"
+#endif
+```
+
+### Usage Example
+This section provides some pseudo code to show what core TensorFlow and plugin's code may look like.
+
+#### Core TensorFlow
+* **ProfileOptions support**
+
+  To enable [ProfileOptions](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/profiler/profiler_options.proto#L6) support for `PluggableProfiler`, we add a new device type `PLUGGABLE_DEVICE` in the `enum DeviceType` field.
+  ```c++
+  // Next ID: 11
+  message ProfileOptions {
+    // Some default values of option are not proto3 default values. Use this version
+    // to determine if we should use default option value instead of proto3
+    // default value.
+    uint32 version = 5;
+  
+    enum DeviceType {
+      UNSPECIFIED = 0;
+      CPU = 1;
+      GPU = 2;
+      TPU = 3;
+      PLUGGABLE_DEVICE = 4;
+    }
+      // Device type to profile/trace: (version >= 1)
+    // DeviceType::UNSPECIFIED: All registered device profiler will be enabled.
+    // DeviceType::CPU: only CPU will be profiled.
+    // DeviceType::GPU: only CPU/GPU will be profiled.
+    // DeviceType::TPU: only CPU/TPU will be profiled.
+    // DeviceType::PLUGGABLE_DEVICE: all pluggable devices with profilers enabled will be profiled. 
```

Comment:
only CPU and pluggable devices with profilers enabled will be profiled

Reply:
Done